### PR TITLE
feat: add structuring executor repositories best practices

### PR DESCRIPTION
**Pull request by tadejsv**
# TLDR

This PR sets up new best practices for importing python modules via the `py_modules` argument:
- For single-file executors, same as before: just list the file in `py_modules`
- For multiple-file executors: organize as a package, put all the files inside a folder (`executor/`) inside your repository, in `py_modules` import only `executor/__init__.py` - see the [addition to Executor cookbook](https://github.com/jina-ai/jina/blob/fix-import/.github/2.0/cookbooks/Executor.md#structure-of-the-repository) in this PR. This is a vast simplification of what we (unofficially) recommended and supported before - listing multiple files in `py_modules` (where order matters), and then patching python's import system to load sub-packages.

It also cleans up some things we added before (like #2953) to support a way of importing that was an anti-pattern in python, and is currently not used and not needed. 

In 3.0, we should go even further and change `py_modules` to `py_module` - there is only a need to import a single file (right now I added a warning if multiple files are imported)

# Background

The reason we were doing things "wrong" before was due to mis-understanding of python's import system. We did everything to try and enable absolute imports - but that only works if we are running the script that is importing as a main file (`python main_file.py`). So we wante to be able to do this
```python
from helper import foo
```
instead of
```python
from .helper import foo
```
I can only speculate what the reasons were - probably because it looks simpler, and works if you execute the main module directly (which also works with relative imports, but you need to have a package directory, more on that later).

When that file is imported as a module - which is what jina does, through `importlib`, absolute imports do not work. And using absolute imports together with `importlib` has other serious drawbacks:
- no way to import whole packages easily - every single module that you use has to be explicitly imported
- suddenly, order of importing matters (with files structured in a package and using relative imports, python takes care of importing magically)

 So we resorted to esoteric patches on python's import system (see #2953) and complicated instructions on how to import files, like this one https://github.com/jina-ai/jina/blob/88d8a6fb0614b87b10c1e3ded3cd3425548874a8/jina/flow/base.py#L197-L198

None of this is needed when using normal Python, so it led to confusion - probably nobody knew exactly how Jina's import system works. Even power users were completely lost (see #1546), and our recommendations were misleading. If we look at tests that we have for this, the confusion becomes obvious - for example here
https://github.com/jina-ai/jina/blob/53e4607c0cabbe5761333ed26c84c0a1b47dc98a/tests/integration/issues/github_1546/test_pymodules_import.py#L35
the statement is not true (it can be before/after, because the example is using relative imports), and here
https://github.com/jina-ai/jina/blob/53e4607c0cabbe5761333ed26c84c0a1b47dc98a/tests/integration/issues/github_1546/good1/custom_crafter.py#L2

the test only worked because the test environment was polluted - if you run the test on its own it will fail, the import should be `from helper import helper_function`.

While I think some people knew that importing as a package (by importing `__init__.py`) works well, there was still some confusion. For example, we did state that this was the recommended solution in some places in the code (but this knowledge appears not to have been disseminated widely), but even those recomendations contained some wrong statements. For example this
https://github.com/jina-ai/jina/blob/88d8a6fb0614b87b10c1e3ded3cd3425548874a8/jina/schemas/meta.py#L29-L33

Here it recommends 
- to import `__init__.py` (good), 
    - but also says that core logic should go there (wrong, no need for that - you only need to import the executor there - see below; and this is also out of line with python packaging conventions)
- that you should use relative imports (good),
- that you should put `helper.py` in `py_modules`, and specifically before `__init__.py` (wrong twice: no need to put it at all, and if you put it there in this case the order will not matter).

# How import system in jina works right now

Currently, the imports in Jina are handled by PathImporter. The important part is this https://github.com/jina-ai/jina/blob/55c49447d66b6e05b9242c99ad58c56a6d199126/jina/importer.py#L180-L201

If we examine this, we see that jina makes an attempt to build some kind of a "general" import system:
- it tries to respect the module names passed in
- if there is conflict with existing packages, tries to rename them

There are some problems with the code above in itself:
- in some cases, the name used to load the module in `spec_from_file_location` (always `'jinahub'`) and name that the module is registered to `sys.modules` to do not match. The effect of this is that if the module imported is an `__init__.py` file - so a package, the relative imports inside this package will not work. This could happen if you imported two `__init__.py` modules - although then you would have other issues as well
- the use of `'jinahub'` as the name of the new module is not optimal - important names such as these should be reserved - we might release a proper python package named jinahub in the future (like huggingface released [huggingface_hub](https://github.com/huggingface/huggingface_hub)).
 
These are, however, only minor issues. What's more problematic is the design of the whole system: it tries to be a general import system (like python's `import` statements), yet has vastly less power. There is no way to define an alias (like `from bar import foo as baz`), shadowing is handled clumsily by prepending `jinahub.` to the name (which I think very few people knew about) - which of course only works once. Should we add an option for users to define aliases? No, that would just compound the already high level of complexity.

So it seems to me that trying to build and maintain such a system is a hopeless task. Rather, let's ask ourselves what import system do we really need, and if this can be achieved with tools already at our disposal.

## Bonus: how are executors registered

So what actually is the purpose of these imports? To get the executor, so that when you write `jtype: MyExecutor`, jina knows how to load `MyExecutor` object. But how does importing a python module do that? 

Well, in Python when a module is imported, its code is run. And if your module defines `MyExecutor` (or imports it, same thing), the class (the class *type*) will be instantiated. And because of inheritances, this will trigger the `JAML.register()` method on the class type, and add it to the yaml resolver - which is responsible for loading the objects.

This means that in the module that we import, we must import (or define) the executor class we want to use.

# How should the import sytem in Jina work

To sort all this out, I first wanted to figure out what purpose does jina's import system serve, which should inform how it should be designed.

The imports (`py_modules`) are meant to be used by the executor (pod?) to load the actual executor. Each executor is run in its own runtime - isolated from other processes. What this means is that we only need to import a single "thing", which gets us our executor. And the name we import it as should not matter - we only do the import so that the module core is run, and executor registered. So if done properly, shadowing etc. should never be a concern.

Therefore we don't need to support importing any number of files/modules, we only need a way to somehow import all files related to one executor, period. And it turns out, if we constrain ourselves to using either single-file executors, or executors organized in a python package, we need to only import a single file.

## Solution

So here's the new best practice for imported that I came up with

- For single-file executors, same as before: just list the file in `py_modules`
- For multiple-file executors: organize as a package, put all the files inside a folder (`executor/`) inside your repository, in `py_modules` import only `executor/__init__.py` (and in it import your executor)

First point is clear, but let me explain a bit more for the second point. It enables you to use complex package structures like this

```
├── config.yml
└── executor/
    ├── helper.py
    ├── __init__.py
    ├── my_executor.py
    └── utils/
        ├── __init__.py
        ├── data.py
        └── io.py
```

With your `config.yml` file looking like this

```yaml
jtype: MyExecutor
metas:
  py_modules:
    - executor/__init__.py
```

and `executor/__init__.py` like this

```python
from .my_executor import MyExecutor
```

From a developer's perspective, this offers several benefits. First of all, you don't need to worry about `py_modules` anymore - just include the `__init__.py` there, done. But on top of that, also:
- this is the usual python packaging system - many users of python will already be familiar with it
- the root folder of the repository is not cluttered with python files/folders
- it enables you to import your executor from python console (or simple test files) from the root of your directory - which is your workspace. Just import as `from executor import MyExecutor`
- you can use this easily with pytest. You also import using `form executor import MyExecutor` in `tests/test_*.py` files.

And let me just discuss one alternative here: putting all python files (including `__init__.py`) inside the root repository folder, so no `executor/` folder. This fails on points 2 and 3. For point 3, specifically: if you try `from . import MyExecutor` (or similar versions), you will be greeted with
```
ImportError: attempted relative import with no known parent package
```

The only way to get around this issue is to go one level up in the directory tree - so to do `cd ..` before starting the python shell. But this is a very strange solution - the root of the repository is your **workspace** when developing the executor, you should be able to stay in the workspace for whatever testing you need to do.

### What should change in 3.0

Now that we know that importing a single file (a module or a package through `__init__.py`) is all that is ever needed, let's in fact limit users to that - the extra "freedom" they would have from importing multiple files can only lead to strange errors, which will lead to a frustrating experience. 

Instead, when they try to import a list, let's throw an error telling them that they should look up how to structure multi-file executors, and a link to the cookbook.

This also allows us to simplify our import system. No need for any checking if there is shadowing and so on - let's just pick a unique name (it does not matter what it is), and import it as that.

TODO:

- [x] Rewrite description/name to focus on new best practice of importing
- [x] Rewrite docstrings/documentation to stress that `py_modules` should be a single file, direct to documentation
- [x] Emit a warning if `py_modules` is a list with len > 1
- [ ] Post notice of new best practice on Slack
- [ ] Check compatibility with jinad / add jinad test with only `__init__.py` as import [should break]

### CODE REVIEW

**Review by JoanFM — CHANGES_REQUESTED** [review]
Please, add unittests to test this change.

**Comment by tadejsv** [comment]
@JoanFM I am trying now, but I have discovered that tests are themselves actually broken right now. Namely, they work because the environment (imports) get modified between tests, but fail if you run only a single test. Try this:

```
pytest tests/integration/issues/github_1546/test_pymodules_import.py::test_import_with_abs_namespace_should_pass
```

**Comment by JoanFM** [comment]
@deepankarm can u please check? I think this needs to be properly tested also in the remote setting with `jinad`

**Review by JoanFM — COMMENTED** [review]
Needs to be tested for how it will work with `JinaD` because I think `jinad` does not allow sending `folders` @deepankarm

**Comment by deepankarm** [comment]
> Needs to be tested for how it will work with `JinaD` because I think `jinad` does not allow sending `folders` @deepankarm

There are few existing tests for this - [1](https://github.com/jina-ai/jina/blob/master/tests/distributed/test_workspaces/test_remote_workspaces.py#L31), [2](https://github.com/jina-ai/jina/blob/master/tests/distributed/test_workspaces/test_remote_workspaces.py#L78), [3](https://github.com/jina-ai/jina/blob/master/tests/unit/peapods/pods/test_pods.py#L278). I think, we are good if CI passes.

**Comment by JoanFM** [comment]
@deepankarm but then we will need to add the possibility to send folders to `JinaD` and import from them right? Because I do not think it will work if the files are flattened?

**Comment by deepankarm** [comment]
> @deepankarm but then we will need to add the possibility to send folders to `JinaD` and import from them right? Because I do not think it will work if the files are flattened?

We have another arg - `upload_files` where one can upload all the other files, it accepts a list.

**Comment by tadejsv** [comment]
@deepankarm I think we need a way to upload a folder - when importing `__init__.py` (something jina supports, and is actually the recommended way) python imports the whole package (folder), so upload should mimic that

**Review by nan-wang — CHANGES_REQUESTED** [review]
minor comments

File: .github/2.0/cookbooks/Executor.md [hunk 1]
```diff
@@ -465,6 +468,101 @@ hello world
...
+
+This way of structuring your files has several advantages:
+- it is the standard way of creating Python packages, so you may already be familiar with it
+- it avoids cluttering the root of your repository with many python files
+- it enables you to try out your executor from the console, i.e. you can do things like this
+    ```python
+    >>> from executor import MyExecutor
+    >>> my_exec = MyExecutor()
+    ```
+- you can easily use it with test automation frameworks, like Pytest. Just remember to import objects from the `executor` package in the test files, e.g. you would use `from executor import MyExecutor` inside `tests/test_unit.py`.
```
[the remark this answers is not in the document]
Comment (tadejsv) [reply .github/2.0/cookbooks/Executor.md]: Agree, removed

**Comment by tadejsv** [comment]
@JoanFM @deepankarm So I think the only issue here is the JinaD incompatibility (when using the executor "locally"). I have opened an issue for this (#3221), and have added a disclaimer to the cookbook about this - and the current documentation also contains some warning about this, see https://github.com/jina-ai/jina/blob/30b441cdf17176045f53b55f70891dea64c92196/jina/parsers/peapods/runtimes/distributed.py#L32-L33

Is there anything else that should be addressed here?

**Review by JoanFM — APPROVED** [review]
